### ISSUE TITLE
Chp.3 Improving the way neural networks learn

### ISSUE THREAD

**Issue by hysic** [hysic/NNDL_ReadingNotes#3]
## 反常识卡
* 反常识：对于前面介绍的二次损失函数（quadratic cost function），如果预测值离实际值很远，损失函数C值反而下降得很慢。
* 原因：二次损失函数C相对于w和b的梯度公式（见下）中都存在σ′(z)一项，这一项在接近0或1的时候都很小，导致学习速率变慢（learning slowdown），这也被称为neuron saturation。
![image](https://user-images.githubusercontent.com/5415646/26919823-8b6cb086-4c69-11e7-87a0-2748ad1cb0de.png)
* 解决方法：将上述公式中的σ′(z)一项去掉，根据梯度反推损失函数，即可得到交叉熵损失函数（cross-entropy cost function）。
* 出处：<http://neuralnetworksanddeeplearning.com/chap3.html#the_cross-entropy_cost_function>

## 术语卡
* 交叉熵损失函数（cross-entropy cost function）
* 定义：单个输出神经元：
![image](https://user-images.githubusercontent.com/5415646/26920077-607e9212-4c6a-11e7-88f3-bee426c47c17.png)
多个输出神经元：
![image](https://user-images.githubusercontent.com/5415646/26920120-765fe3d8-4c6a-11e7-8e6f-73aa1422b828.png)
其中a为模型输出，y为真实输出。
* 印象：这个函数可以作为损失函数，原因有二：①函数非负；②a趋近于y时，C趋近于0。此外这个定义也满足上一章中所说cost function想要应用BP算法所要满足的[两个条件](https://github.com/hysic/NNDL_ReadingNotes/issues/2#issue-232494563)。
* 优点：与二次损失函数相比，交叉熵损失函数收敛得更快，通常也可以得到更好的结果。
* 出处：<http://neuralnetworksanddeeplearning.com/chap3.html#the_cross-entropy_cost_function>

## 任意卡
* 如果输出神经元是线性神经元，那么应该使用二次损失函数，这种情况没有学习速率变慢的问题。

-----
根据[吴恩达的DeepLearning课程](https://mooc.study.163.com/learn/deeplearning_ai-2001281002?tid=2001392029#/learn/content?type=detail&id=2001701009&cid=2001694012)，对于LR回归使用cross-entropy损失函数，而不用二次损失函数，是因为二次损失函数对LR来说，在优化时是非凸的（nnon-convex），会得到多个局部最优解。

**Comment by hysic** [comment]
## 术语卡
* Softmax layer （替代最后一层的sigmoid layer）
* 定义：softmax激活函数的公式如下：
![image](https://user-images.githubusercontent.com/5415646/26920772-a3d14454-4c6c-11e7-88f7-76df851a0f71.png)
所有输出神经元的softmax激活函数的总和为1，所以可以看作是一种概率分布。
* softmax也可以解决学习速率变慢的问题，但需要结合对数似然（log-likelihood）损失函数，其定义如下：
![image](https://user-images.githubusercontent.com/5415646/26920913-0fd169d6-4c6d-11e7-9f72-ecc246764e91.png)
相应的梯度公式为：
![image](https://user-images.githubusercontent.com/5415646/26920942-1e878ea6-4c6d-11e7-8d17-c512a56a571b.png)
* softmax输出层 + 对数似然损失函数 ≈ sigmoid输出层 + 交叉熵损失函数
* softmax名称的来历：对max（或argmax）函数的平滑化（smooth）或可微分化（<https://www.quora.com/Why-is-softmax-activate-function-called-softmax>）

**Comment by hysic** [comment]
## 故事卡

The Nobel prizewinning physicist Enrico Fermi was once asked his opinion of a mathematical model some colleagues had proposed as the solution to an important unsolved physics problem. The model gave excellent agreement with experiment, but Fermi was skeptical. He asked how many free parameters could be set in the model. "Four" was the answer. Fermi replied*: "I remember my friend Johnny von Neumann used to say, with four parameters I can fit an elephant, and with five I can make him wiggle his trunk.".

*The quote comes from a charming article by Freeman Dyson, who is one of the people who proposed the flawed model. A four-parameter elephant may be found [here](https://www.johndcook.com/blog/2011/06/21/how-to-fit-an-elephant/). 


![image](https://user-images.githubusercontent.com/5415646/26921844-dd88db3c-4c6f-11e7-97fa-97787200e38e.png)

**Comment by hysic** [comment]
* 过拟合（overfitting）
![image](https://user-images.githubusercontent.com/5415646/27033610-be64ae76-4fac-11e7-9589-e6579091286b.png)
* 如何发现过拟合 -> 训练集 + 验证集 + 测试集
* 发现了过拟合，如何降低 -> 增大训练数据量；降低神经网络复杂度；正则化。

**Comment by hysic** [comment]
## 正则化（regularization）
* L2正则化
    * 定义：向损失函数中添加正则项，比如对于交叉熵损失函数和二次损失函数来说，添加正则项后的公式都可以写作：
![image](https://user-images.githubusercontent.com/5415646/27258098-be3e697a-5423-11e7-91b6-16ca5b8049b0.png)
其中C_0是原始未正则化的损失函数，λ是正则化参数（λ>0）。
    * 通过调节λ，在最小化原始损失函数和最小化权重之间取折中。
    1. 这种正则化的确可以降低过拟合。从下面两个式子中可以看出，
![image](https://user-images.githubusercontent.com/5415646/27258834-9799de1c-5437-11e7-9f8a-2f7b2c75c4d3.png)
![image](https://user-images.githubusercontent.com/5415646/27258838-a05f9eba-5437-11e7-94ad-86be1e8e5bfc.png)
由于1−ηλ/n因子的存在，w的绝对值相对非正则化要小，所以L2正则化也叫做权重衰减（weight decay）。从代码实际运行结果来看，L2正则化的确降低了过拟合。
        * L2正则化不仅可以降低过拟合，提高验证精度，而且与未正则化相比，更不容易陷入局部最小值，其结果更容易重现。

    2. 正则化降低过拟合的原因
        * 正则化神经网络->权重较小->输入发生较小的随机变化（噪声），输出不会变化较大。
        * 换句话说，正则化神经网络倾向于学习简单的模型（奥卡姆剃刀，Occam's Razor），不容易受数据噪声的影响。
        * 科学偏爱简单模型这一说法并不是必然正确的，作者举了两个例子，在这两个例子中都是更复杂的模型胜利了。第一个是Bethe和Schein关于是否发现新粒子的争论（mesons），第二个是牛顿力学和广义相对论对水星进动的解释。“简单”的定义并不明确。
        * 从经验来看，正则化的神经网络模型更简单，泛化性能也更好，但目前尚无令人信服的理论解释。
        * 100个隐神经元的神经网络具有约80000个参数，用来拟合50000张图像，应该已经严重过拟合了，但实际拟合效果还不错。Yann LeCun将其解释为神经网络具有自正则化效应（self-regularization effect）。
    * L2正则化仅压缩权重w，未压缩偏置b -> 如果b很大，输入的微小变化不会造成输出发生很大变化，所以无需压缩b。此外，b较大容易使神经元饱和，这是我们希望得到的（why？）。

**Comment by hysic** [comment]
## 其他正则化方法
### 1. L1正则化
* L1正则化损失函数及其梯度：
![image](https://user-images.githubusercontent.com/5415646/27507017-1bc840be-58f9-11e7-9f63-126fc36af701.png)
![image](https://user-images.githubusercontent.com/5415646/27507023-413cf98e-58f9-11e7-9eb1-a014f311e5dd.png)
![image](https://user-images.githubusercontent.com/5415646/27507025-47e029e6-58f9-11e7-8dd0-4ac3c3191806.png)
* 与L2正则化对比
    * 相同：使权重变小
    * 不同：权重的缩小速度不同。在L2正则化中，权重的缩小速度与权重大小成正比（ηλ/n * w）；而在L1正则化中，权重的缩小速度是恒定的。因此，对于较大的权重，L2正则化对权重的缩小速度更快；而对于较小的权重，L1正则化对权重的缩小速度更快。
* L1正则化仅选取少数特别重要的权重，其他权重趋于0。
 
### 2.  Dropout
* L1和L2修改损失函数，而dropout修改神经网络。
* 做法：①随机删除一部分隐神经元（比如1/2），然后进行前向传播和反向传播计算梯度，训练一个mini-batch，然后恢复这部分神经元，多次重复上述过程。②在用dropout训练好的神经网络进行预测时，需要用整个网络进行预测，即隐神经元变多，需要将隐神经元向外的权重乘以相同的比例（比如1/2）。
* 降低过拟合的原因：每次dropout相当于训练不同的神经网络，多次dropout相当于对不同神经网络的结果去平均值，由于不同神经网络的过拟合方式各不相同，取平均之后可以降低过拟合。

### 3. 人工增加训练数据量
* 对于图像数据：rotations(旋转), translating(平移), skewing(歪斜), elastic distortions()，可参考[1](https://pdfs.semanticscholar.org/7b1c/c19dec9289c66e7ab45e80e8c42273509ab6.pdf)，[2](https://stackoverflow.com/questions/2446494/skewing-an-image-using-perspective-transforms)
* 对于语音数据：添加背景噪声、加快或放慢语音速度
* 这些方法不一定总会用到：对原始语音数据进行降噪，比添加背景噪声来扩充数据更加有效。

**Comment by hysic** [comment]
![](http://neuralnetworksanddeeplearning.com/images/more_data_comparison.png)
从上图可以看出，50000张图片的SVM要优于1000张图片的神经网络。因此，在比较两个算法的优劣时，一定要注意所使用的训练数据。我们的目的是好算法+好数据。在寻找最优算法时，一定要先看一下有没有更简单的方法可以获取更多更好的数据。
![image](https://user-images.githubusercontent.com/5415646/27728078-89072c94-5db3-11e7-828e-b0d0b117bf89.png)

**Comment by hysic** [comment]
## 权重初始化（Weight Initialization）
* 最开始介绍的初始化方法为：w,b ~ N(0, 1)。这种方法的问题在于，假设有1000个输入神经元，有一半概率等于1，另一半概率等于0，那么对于第一个隐层中的神经元来说，其输入满足N(0, \sqrt{501})，标准差很大，绝对值很容易远大于1，造成神经元饱和，学习速率变慢。
* 解决方法：w ~ N(0, 1/\sqrt{N_in})，b ~ N(0, 1)，这样第一个隐层中的神经元输入满足N(0, \sqrt{3/2})。
* 优点：既可以加快学习速度，也可以提高最终性能。

**Comment by hysic** [comment]
* 如何选择神经网络的超参数 (e.g. learning rate, η; regularization parameter, λ; etc.)?
    * 本章介绍调参的heuristics。
    * 一般策略：尝试使结果优于随机猜测。
        * 简化训练数据（减少数据量，减少标签数量）
        * 简化模型
        * 快速查看训练结果，得到快速反馈（比如每个mini-batch都输出评估精度）
        * 根据在验证集上的精度进行评价
        * 逐步迭代
        * 万事开头难
    * 调节学习速率η
        * 寻找使训练集损失函数在前几个epoch减小的最大η，即临界值。
**可变学习速率：e.g.** 开始取一个较大的学习速率，如果验证精度连续n次没有增大，则将学习速率减半，当学习速率减小到初始值的比如1/256时停止。
    * 确定训练epoch数
        * early stropping: 对于MNIST，在初期，如果连续10次的验证精度都没有提高，可以终止训练；在后期，可以将连续的次数增多。对于其他数据集，early stopping的次数需要重新实验。
    * 调节正则化参数λ 
        * 首先令λ =0，找到最佳学习速率η；固定η，将λ 初始化为1.0，首先按10倍增减，以观察验证精度的变化，然后再微调；找到最佳λ 之后，再次调节学习速率η。
    * 调节Mini-batch size
        * size太小，无法充分利用矩阵运算库；size太大，更新权重的速度又太慢。
        * size独立于其他超参数。
        * 寻找验证精度随时间增长最快所对应的size。
    * 自动化调参
        * 网格搜索（grid search）
        * 贝叶斯方法（<http://papers.nips.cc/paper/4522-practical-bayesian-optimization-of-machine-learning-algorithms.pdf>）
    * 超参数优化问题远没有完全解决。

**Comment by hysic** [comment]
## 随机梯度下降（SGD）的变体
* Hessian方法
    * ![image](https://user-images.githubusercontent.com/5415646/27811692-8f8a388c-609b-11e7-9ff7-501bd462858d.png)
    * 优点：收敛快；缺点：计算量大。
* 基于动量的梯度下降
    * ![image](https://user-images.githubusercontent.com/5415646/27813496-f988ad56-60a8-11e7-920e-0b3702cd109d.png)

## 其他类型的神经元/激活函数（除了Sigmoid）
* tanh神经元
    * ![image](https://user-images.githubusercontent.com/5415646/27813503-0a7516a4-60a9-11e7-8a63-ee51f4c2f2dd.png)
    * 有时比Sigmoid表现好。
* ReLU神经元（rectified linear unit）
    * ![image](https://user-images.githubusercontent.com/5415646/27813622-bebea094-60a9-11e7-88d2-df2832783970.png)
    * 其何时有效已经为何有效并不明确。
